sympy/functions/combinatorial/numbers.py

```diff
@@ -736,36 +770,72 @@ class euler(Function):
     """
 
     @classmethod
-    def eval(cls, m):
-        if m.is_odd:
-            return S.Zero
-        if m.is_Integer and m.is_nonnegative:
-            from mpmath import mp
-            m = m._to_mpmath(mp.prec)
-            res = mp.eulernum(m, exact=True)
-            return Integer(res)
+    def eval(cls, m, sym=None):
+        if m.is_Number:
+            if m.is_Integer and m.is_nonnegative:
+                # Euler numbers
+                if sym is None:
+                    if m.is_odd:
+                        return S.Zero
+                    from mpmath import mp
+                    m = m._to_mpmath(mp.prec)
+                    res = mp.eulernum(m, exact=True)
+                    return Integer(res)
+                # Euler polynomial: numerical evaluation
+                elif sym.is_Float or (sym.is_number and sym.is_complex \
+                        and all([a.is_Float for a in sym.as_real_imag()])):
```

Comment:
Would it suffice to use `pure_complex` here?

Reply:
yes!  and doing so, we can improve the logic.  Thanks for showing me `pure_complex`.